### ISSUE TITLE
Argument constraint That (or Ignored, or _), when nested deeper in A.CallTo, misreports what's being matched

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#1177]
Found while investigating  [FakeItEasy - How to verify nested arguments value C#](https://stackoverflow.com/questions/45411178/fakeiteasy-how-to-verify-nested-arguments-value-c-sharp). 

Consider this test:

```c#
public class NestedThat
{
    public class Dto
    {
        public string Id;
    }

    public interface IService
    {
        void DoIt(Dto dto);
    }

    [Test]
    public void Test()
    {
        var fake = A.Fake<IService>();
        fake.DoIt(new Dto { Id = "some id" });

        A.CallTo(() => fake.DoIt(new Dto { Id = A<string>._ })).MustHaveHappened();
    }
}
```

The author would expect a passing test, but it fails, which isn't horrible, since the use of `A<string>._` inside the `Dto constructor` is illegal. However, the failure does not lead the reader to the fact that the use of `._` (and I'm sure any variant of `That`) is bad. Here's the output:

```
Test 'FakeItEasyPlay.NestedThat.Test' failed: FakeItEasy.ExpectationException : 

  Assertion failed for the following call:
    FakeItEasyPlay.NestedThat+IService.DoIt(dto: <Ignored>)
  Expected to find it at least once but found it #0 times among the calls:
    1: FakeItEasyPlay.NestedThat+IService.DoIt(dto: FakeItEasyPlay.NestedThat+Dto)

	at FakeItEasy.Core.FakeAsserter.AssertWasCalled(Func`2 callPredicate, String callDescription, Func`2 repeatPredicate, String repeatDescription)
	at FakeItEasy.Configuration.RuleBuilder.MustHaveHappened(Repeated repeatConstraint)
	at FakeItEasy.AssertConfigurationExtensions.MustHaveHappened(IAssertConfiguration configuration)
	NestedThat.cs(19,0): at FakeItEasyPlay.NestedThat.Test()
```

In addition to not highlighting the illegal `._`, we are misreporting the expected argument to `DoIt` as being equivalent to `A<Dto>._`.

It may be difficult (impossible) to throw correct error in all cases, but in this one, where we're capturing  a `That` constraint to match against a parameter of a different type, I'd hope we'd be able to do _something_.

**Comment by thomaslevesque** [comment]
> where we're capturing a That constraint to match against a parameter of a different type

I'm a bit lost... in this case, you're not using `That` at all, and I don't see where you use a "different type". I think the problem exists for any argument constraint, not just `That`.

Anyway, it's true that the error is misleading.

We could transform the expression so that argument constraints on properties of the arguments are evaluated as well. For instance, `new Dto { Id = A<string>._ }` would become `A<Dto>.That.IsNotNull()` (because no property is actually constrained), and `new Dto { Id = A<string>.That.Matches(id => predicate(id))}` would become `A<Dto>.That.Matches(dto => predicate(dto.Id))`. But it would be difficult to get right... For instance, should `new Dto { Id = "a", Name = "Foo" }` be a match for the constraint `new Dto { Id = A<string>._ }`?

A simpler solution would be to just not allow argument constraints in object initializers...

**Comment by blairconrad** [comment]
Oh, `_` is `That`.
And the type mismatch is `string` vs. `Dto`.

I see your workarounds, but it's not just argument initializers. I used one for convenience. The original code used a constructor to make the `Dto` analogue.

**Comment by thomaslevesque** [comment]
> I see your workarounds, but it's not just argument initializers. I used one for convenience. The original code used a constructor to make the Dto analogue.

Ah, yes. Well, we could allow argument constraints only at the "root" of the argument expression. I don't think it's possible to support nested constraints, because the semantic is ambiguous, and anyway we have no way to know what was passed as a constructor argument.

**Comment by blairconrad** [comment]
> we could allow argument constraints only at the "root" of the argument expression

Yeah. I think this is the way.

**Comment by thomaslevesque** [comment]
It would be nice to update the analyzer to detect this too.

**Comment by blairconrad** [comment]
This change has been released as part of [FakeItEasy 4.1.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/4.1.0).
